fendermint/testing/materializer/src/docker/node.rs

```diff
@@ -0,0 +1,566 @@
+// Copyright 2022-2024 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use std::{
+    collections::{BTreeMap, HashMap},
+    os::unix::fs::MetadataExt,
+    path::{Path, PathBuf},
+};
+
+use anyhow::{anyhow, Context};
+use bollard::{
+    container::{Config, CreateContainerOptions, RemoveContainerOptions},
+    service::{HostConfig, PortBinding},
+    Docker,
+};
+use ethers::types::H160;
+use lazy_static::lazy_static;
+
+use super::{
+    container::DockerContainer, DockerConstruct, DockerMaterials, DockerNetwork, DockerPortRange,
+    DockerWithDropHandle,
+};
+use crate::{
+    docker::DOCKER_ENTRY_FILE_NAME,
+    materializer::{NodeConfig, TargetConfig},
+    materials::export_file,
+    NodeName, ResourceHash,
+};
+
+// TODO: Add these to the materializer.
+const COMETBFT_IMAGE: &str = "cometbft/cometbft:v0.37.x";
+const FENDERMINT_IMAGE: &str = "fendermint:latest";
+
+const STATIC_ENV: &str = "static.env";
+const DYNAMIC_ENV: &str = "dynamic.env";
+
+const RESOLVER_P2P_PORT: u32 = 26655;
+const COMETBFT_P2P_PORT: u32 = 26656;
+const COMETBFT_RPC_PORT: u32 = 26657;
+const FENDERMINT_ABCI_PORT: u32 = 26658;
+const ETHAPI_RPC_PORT: u32 = 8445;
+
+lazy_static! {
+    static ref STATIC_ENV_PATH: String = format!("/opt/docker/{STATIC_ENV}");
+    static ref DYNAMIC_ENV_PATH: String = format!("/opt/docker/{DYNAMIC_ENV}");
+    static ref DOCKER_ENTRY_PATH: String = format!("/opt/docker/{DOCKER_ENTRY_FILE_NAME}");
+}
+
+type EnvVars = BTreeMap<&'static str, String>;
+type Volumes = Vec<(PathBuf, &'static str)>;
+
+macro_rules! env_vars {
+    ( $($key:literal => $value:expr),* $(,)? ) => {
+        BTreeMap::from([ $( ($key, $value.to_string()) ),* ])
+    };
+}
+
+/// A Node consists of multiple docker containers.
+pub struct DockerNode {
+    /// Logical name of the node in the subnet hierarchy.
+    node_name: NodeName,
+    fendermint: DockerContainer,
+    cometbft: DockerContainer,
+    ethapi: Option<DockerContainer>,
+    port_range: DockerPortRange,
+    path: PathBuf,
```

Comment:
is this the path on the host machine which is shared as volume with the docker contains?

Reply:
I'll add a comment; yes, it's the path on the host machine to all the files and directories for the containers. Not all are volumes, but most are. I added an example to the description of https://github.com/consensus-shipyard/ipc/pull/744